### PR TITLE
Interception improvements (merging implementations + acceleration)

### LINKED ISSUE #1799: BallFilter: Determine ball acceleration
### Description of the task
The `Ball` class has an acceleration field that is used to determine future ball states. However, it is initialized to 0 because the `BallFilter` will only determine position and velocity. We should improve the `BallFilter` to determine the acceleration that the ball is experiencing it so we can improve an accelerating ball, from e.g. friction, spin, etc.

### Acceptance criteria

- [ ] Improve the `BallFilter` to determine acceleration
- [ ] Use acceleration in `Ball`
- [ ] Test the filter

### PR DESCRIPTION
**Pull request by amrm3lm**
### Description

changes to interception:
- merged two implementations we had for interception: we know use gradient descent until the ball is close or slow enough to chase
- added a constant acceleration due to rolling friction into `ball_filter` , which is also a dynamic parameter

### Testing Done
Added tests to `ball_filter` and `ball`, also checked simulated tests to see improvement. 
One example is here: https://imgur.com/a/Eimbe0j

### Resolved Issues
resolves #1799 , #2157 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/world/ball.cpp [hunk 1]
```diff
@@ -55,11 +57,30 @@ Vector Ball::acceleration() const
...
+    // since acceleration is due to friction, final velocity cannot be in the opposite
+    // direction to initial velocity. if that happens, we find a new velocity using a more
+    // realistic Duration
+    Duration effective_duration(duration_in_future);
+
+
+    if (acceleration_.length() > 0 &&
+        almostEqual(future_velocity.rotate(Angle::half()).orientation().toRadians(),
+                    current_state_.velocity().orientation().toRadians(), FIXED_EPSILON,
+                    ULPS_EPSILON_TEN))
```
Comment (jonathanlew) [inline src/software/world/ball.cpp]: this is pretty confusing, can you explain and add a comment

File: src/software/sensor_fusion/filter/ball_filter.h [hunk 2]
```diff
@@ -43,11 +43,13 @@ class BallFilter
     static constexpr double MAX_BUFFER_SIZE_VELOCITY_MAGNITUDE = 4.0;
     // The extra amount beyond the ball's max speed that we treat ball detections as valid
     static constexpr double MAX_ACCEPTABLE_BALL_SPEED_BUFFER = 2.0;
+    // the speed after which we start to experience rolling friction
+    static constexpr double BALL_MIN_SPEED_FOR_ROLLING_ACCELERATION = 0.2;
```
Comment (jonathanlew) [inline src/software/sensor_fusion/filter/ball_filter.h:47]: Why is this necessary? Why can't we always apply acceleration?

**Review by jonathanlew — COMMENTED** [review]
looks good, got a few comments (this pr should target robocup_2021)

File: src/software/world/ball.cpp (hunk 1 above)
Comment (amrm3lm) [reply src/software/world/ball.cpp]: its checking if the final velocity (after friction has been applied for the whole duration) is in the opposite direction to the velocity the ball is currently in. We check this because friction will only be applied until the ball completely stops. So if the final velocity is in the opposite direction we need to use a different duration to find position. 

there is a comment above the if statement with explanation, i moved this check into its own variable to make it clearer

File: src/software/sensor_fusion/filter/ball_filter.h (hunk 2 above)
Comment (amrm3lm) [reply src/software/sensor_fusion/filter/ball_filter.h:47]: friction will only be applied if the ball is in motion, but to account for vision not being completely accurate theres a +- 0.2 where we'll assume the ball is stationary and thus has no friction applied on it

File: src/software/ai/hl/stp/play/defense_play_test.cpp [hunk 3]
```diff
@@ -155,14 +155,14 @@ TEST_F(DefensePlayTest, test_defense_play_close_to_net)
             // Wait for all robots to come to a halt
             robotHalt(world_ptr, yield);
             // Attacker in front of enemy with the ball
-            Rectangle attacker_rect(Point(-3.5, 1), Point(-2.3, 0));
-            robotInPolygon(1, attacker_rect, world_ptr, yield);
+            //            Rectangle attacker_rect(Point(-3.5, 1), Point(-2.3, 0));
+            //            robotInPolygon(4, attacker_rect, world_ptr, yield);
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/play/defense_play_test.cpp]: nit: just delete this

**Review by jonathanlew — DISMISSED** [review]

**Comment by amrm3lm** [comment]
Closing this for the time being. Was used for robocup, but following that we need to properly assess the requirements of acceleration and need a better way of comparing interception algorithms.
